### PR TITLE
feat: wallet FFI cucumber

### PR DESCRIPTION
**Pull request by Cifko**
Description
---
Added some ffi steps, with some ffi calls and structures for better handling the ffi.
The biggest problem right now is that I don't know how to create extern "C" closure. So all ffi wallets share the callbacks right now.

How Has This Been Tested?
---
Running the FFI feature in cucumber. So far so good.

### CODE REVIEW

File: base_layer/wallet_ffi/src/lib.rs [hunk 1]
```diff
@@ -4619,6 +4619,65 @@ pub unsafe extern "C" fn comms_list_connected_public_keys(
     }
 }
 
+/// Gets the length of the public keys vector
+///
+/// ## Arguments
+/// `*const TariPublicKeys` - Pointer to emoji set
```
Comment (SWvheerden) [inline base_layer/wallet_ffi/src/lib.rs]: public keys?

File: base_layer/wallet_ffi/src/lib.rs [hunk 2]
```diff
@@ -4619,6 +4619,65 @@ pub unsafe extern "C" fn comms_list_connected_public_keys(
...
+    ptr::swap(error_out, &mut error as *mut c_int);
+    if public_keys.is_null() {
+        error = LibWalletError::from(InterfaceError::NullError("public_keys".to_string())).code;
+        ptr::swap(error_out, &mut error as *mut c_int);
+        return 0;
+    }
+    (*public_keys).0.len() as c_uint
+}
+
+/// Gets a ByteVector at position in a EmojiSet
```
Comment (SWvheerden) [inline base_layer/wallet_ffi/src/lib.rs:4742]: public keys?

File: integration_tests/tests/cucumber.rs [hunk 3]
```diff
@@ -599,14 +633,15 @@ async fn send_amount_from_source_wallet_to_dest_wallet_without_broadcast(
...
 
-    let mut dest_client = create_wallet_client(world, dest_wallet.clone()).await.unwrap();
-    let dest_wallet_pubkey = dest_client
-        .identify(identity_req)
-        .await
-        .unwrap()
-        .into_inner()
-        .public_key
-        .to_hex();
+    let dest_wallet_pubkey = world.get_wallet_address(&dest_wallet).await.unwrap();
```
Comment (SWvheerden) [inline integration_tests/tests/cucumber.rs]: its not a pub key, its a tari address

File: integration_tests/tests/cucumber.rs [hunk 4]
```diff
@@ -599,14 +633,15 @@ async fn send_amount_from_source_wallet_to_dest_wallet_without_broadcast(
...
+    // let mut dest_client = create_wallet_client(world, dest_wallet.clone()).await.unwrap();
+    // let dest_wallet_pubkey = dest_client
+    //     .identify(identity_req)
+    //     .await
+    //     .unwrap()
+    //     .into_inner()
+    //     .public_key
+    //     .to_hex();
```
Comment (SWvheerden) [inline integration_tests/tests/cucumber.rs]: ?

File: integration_tests/tests/cucumber.rs [hunk 5]
```diff
@@ -674,14 +716,15 @@ async fn send_one_sided_transaction_from_source_wallet_to_dest_wallt(
...
```
Comment (SWvheerden) [inline integration_tests/tests/cucumber.rs]: ?

File: integration_tests/tests/cucumber.rs [hunk 6]
```diff
@@ -674,14 +716,15 @@ async fn send_one_sided_transaction_from_source_wallet_to_dest_wallt(
...
 
-    let mut dest_client = create_wallet_client(world, dest_wallet.clone()).await.unwrap();
-    let dest_wallet_pubkey = dest_client
-        .identify(identity_req)
-        .await
-        .unwrap()
-        .into_inner()
-        .public_key
-        .to_hex();
+    let dest_wallet_pubkey = world.get_wallet_address(&dest_wallet).await.unwrap();
```
Comment (SWvheerden) [inline integration_tests/tests/cucumber.rs]: tari address

File: integration_tests/tests/cucumber.rs [hunk 7]
```diff
@@ -1201,18 +1244,19 @@ async fn send_num_transactions_to_wallets_at_fee(
...
-        .await
-        .unwrap()
-        .into_inner();
+    let dest_wallet_pubkey = world.get_wallet_address(&receiver_wallet).await.unwrap();
+    // let mut receiver_wallet_client = create_wallet_client(world, receiver_wallet.clone()).await.unwrap();
+    // let receiver_wallet_identity_res = receiver_wallet_client
+    //     .identify(GetIdentityRequest {})
+    //     .await
+    //     .unwrap()
+    //     .into_inner();
```
Comment (SWvheerden) [inline integration_tests/tests/cucumber.rs]: ?

File: integration_tests/tests/cucumber.rs [hunk 8]
```diff
@@ -1201,18 +1244,19 @@ async fn send_num_transactions_to_wallets_at_fee(
 ) {
     let mut sender_wallet_client = create_wallet_client(world, sender_wallet.clone()).await.unwrap();
 
-    let mut receiver_wallet_client = create_wallet_client(world, receiver_wallet.clone()).await.unwrap();
-    let receiver_wallet_identity_res = receiver_wallet_client
-        .identify(GetIdentityRequest {})
-        .await
-        .unwrap()
-        .into_inner();
+    let dest_wallet_pubkey = world.get_wallet_address(&receiver_wallet).await.unwrap();
```
Comment (SWvheerden) [inline integration_tests/tests/cucumber.rs]: address

**Review by SWvheerden — CHANGES_REQUESTED** [review]
This PR we use Public key for the wallet addresses, this is not correct. 

Wallets use Tari addresses, and BaseNodes use public keys. 

Although, a TariAddress uses a public key inside of it, it is not a public key and I think we should ensure that the code reflects this. This PR currently uses public keys

File: base_layer/wallet_ffi/src/lib.rs (hunk 1 above)
Comment (Cifko) [reply base_layer/wallet_ffi/src/lib.rs]: That reminds me that there are more typos like this. But going over ~176 functions is not part of this PR.

File: base_layer/wallet_ffi/src/lib.rs (hunk 2 above)
Comment (Cifko) [reply base_layer/wallet_ffi/src/lib.rs:4742]: These are indeed public keys. It's the set of connected base nodes.

**Comment by Cifko** [comment]
> 

This PR is almost complete. But it needs to pass the integration test and for that I need to wait for #5077 so I can run proper tests.

**Comment by jorgeantonio21** [comment]
PR #5077 addresses the issue with public keys vs tari addresses

**Pull request merged**